### PR TITLE
feat(reborn): enable NEAR AI MCP from Reborn env config

### LINKED ISSUE #4700: Enable NEAR AI MCP automatically when NEAR AI credentials are configured
## Summary

When local IronClaw/Reborn is configured with both `NEARAI_BASE_URL` and `NEARAI_API_KEY`, the NEAR AI MCP integration should be enabled automatically instead of requiring an extra manual setup step.

## Observed Behavior

During local setup/testing, NEAR AI provider credentials can already be present in `.env`, but NEAR AI MCP is not automatically enabled from that existing configuration.

## Expected Behavior

If the runtime sees valid NEAR AI configuration, for example:

```env
NEARAI_BASE_URL=...
NEARAI_API_KEY=...
```

then NEAR AI MCP should be available/enabled by default, or the setup flow should clearly activate it as part of provider setup.

## Why This Matters

A user who has already configured NEAR AI credentials expects NEAR AI-backed capabilities to work without discovering and enabling a separate MCP integration manually. This is especially important for local testing and first-run setup ergonomics.

## Acceptance Criteria

- NEAR AI MCP is enabled automatically when `NEARAI_BASE_URL` and `NEARAI_API_KEY` are configured and valid.
- The behavior is documented in the relevant setup/Reborn docs.
- If automatic enablement is unsafe or not possible, the WebUI/setup flow should present an explicit, actionable prompt to enable it.
- Add focused coverage for the config-detection path.

### PR DESCRIPTION
**Pull request by think-in-universe**
Fixes #4700

## Summary
- Detect `NEARAI_BASE_URL` plus `NEARAI_API_KEY` during Reborn local-dev startup.
- Seed the NEAR AI API key through Reborn product-auth manual-token storage, updating an existing NEAR AI account when present.
- Install and statically activate the bundled `nearai` MCP extension when auto-activatable, while preserving explicit disabled state.
- Document the Reborn local-dev auto-bootstrap behavior and update parity/env docs.

## Tests
- `cargo +1.92.0 fmt`
- `git diff --check`
- `cargo +1.92.0 test -p ironclaw_reborn_composition local_dev_nearai_mcp_auto_bootstraps_from_env_credentials`

Note: the targeted test passes with two pre-existing warnings in `crates/ironclaw_reborn_composition/src/runtime.rs` unrelated to this change.

### CODE REVIEW

File: crates/ironclaw_reborn_composition/src/nearai_mcp.rs [hunk 1]
```diff
@@ -15,6 +15,62 @@ pub(crate) fn nearai_mcp_endpoint_from_env() -> Result<NearAiMcpEndpoint, String
...
+        .map(|value| value.trim().to_string())
+        .filter(|value| !value.is_empty())?;
+    let api_key = std::env::var("NEARAI_API_KEY")
+        .ok()
+        .map(|value| value.trim().to_string())
+        .filter(|value| !value.is_empty())?;
+    Some((configured_base, secrecy::SecretString::from(api_key)))
+}
+
+#[cfg(test)]
```
Comment (serrrfirat) [inline crates/ironclaw_reborn_composition/src/nearai_mcp.rs]: This override is process-global mutable state. Because `build_local_dev()` now always consults `nearai_mcp_env_credentials()`, any other local-dev test that runs while this guard is live will also see the synthetic credentials and unexpectedly auto-bootstrap NEAR AI MCP. Please switch this to an injected test seam or a scoped env helper that serializes the affected tests.

File: crates/ironclaw_reborn_composition/src/factory.rs [hunk 2]
```diff
@@ -320,6 +322,134 @@ where
         })
 }
 
+async fn bootstrap_local_dev_nearai_mcp_from_env(
```
Comment (serrrfirat) [inline crates/ironclaw_reborn_composition/src/factory.rs]: This whole NEAR-AI-specific bootstrap state machine now lives in the 4.6k-line composition root and open-codes manual-token setup, credential rotation, lifecycle projection, install, and activation. That gives `factory.rs` a second product-auth / extension-lifecycle path that will drift the next time either policy changes. Can this move behind a dedicated local-dev bootstrap helper that reuses the canonical credential-setup and lifecycle abstractions instead of wiring the protocol here?

**Review by serrrfirat — COMMENTED** [review]
Two issues from the hourly sweep: one real test-isolation problem, and one structural concern in the new local-dev bootstrap path.

**Review by abbyshekit — COMMENTED** [review]
## Code review — feat(reborn): enable NEAR AI MCP from Reborn env config

Multi-agent review (security · bugs · performance · tests · conventions · thermo-nuclear) at `71dccdca`. Diff-only mode.
> Independent review of an open PR.

**7 findings** — 4 Medium, 3 Low. Posted as a comment (advisory). Confidence ≥ 50, deduplicated across reviewers.

| Sev | Conf | Reviewer | Location | Finding |
|---|---|---|---|---|
| Medium | 80% | tests | `factory.rs:376` | update-existing-account path (with_update_binding) and re-bootstrap idempotency are untested |
| Medium | 80% | thermo | `factory.rs:354` | Manual-token seeding reimplemented inline instead of reusing the canonical ProductAuthExtensionCredentialSetup helper |
| Medium | 78% | thermo | `factory.rs:414` | Duplicated activate() arm across Discovered/Installed phases — collapse to install-if-needed + single activate |
| Medium | 70% | tests | `factory.rs:438` | disabled-state-preserved branch has no test despite being a documented guarantee |
| Low | 70% | thermo | `factory.rs:325` | 126-line NEAR-AI-MCP bootstrap added to an already-4671-line factory.rs; belongs in the nearai_mcp sibling module |
| Low | 60% | tests | `factory.rs:333` | invalid NEARAI_BASE_URL -> build failure error path is not exercised end-to-end |
| Low | 55% | tests | `nearai_mcp.rs:18` | nearai_mcp_env_credentials one-var-missing / empty / whitespace branches are unexercised |

---
<sub>Generated by near-ai-code-review (5 parallel reviewer agents + intent analysis). Diff-only; confidence ≥ 50; ≤ 15 inline comments.</sub>

File: crates/ironclaw_reborn_composition/src/factory.rs [hunk 3]
```diff
@@ -320,6 +322,134 @@ where
...
+        provider,
+        ironclaw_auth::CredentialAccountLabel::new("NEAR AI API key").map_err(|error| {
+            RebornBuildError::InvalidConfig {
+                reason: format!("NEAR AI MCP credential label is invalid: {error}"),
+            }
+        })?,
+        ironclaw_auth::AuthContinuationRef::SetupOnly,
+        chrono::Utc::now() + chrono::Duration::minutes(5),
+    );
+    if let Some(account) = existing {
```
Comment (abbyshekit) [inline crates/ironclaw_reborn_composition/src/factory.rs]: **[Medium · 80% confidence] update-existing-account path (with_update_binding) and re-bootstrap idempotency are untested**

_tests reviewer_

`bootstrap_local_dev_nearai_mcp_from_env` runs on every local-dev startup. When a NEAR AI product-auth account already exists it takes the `with_update_binding` branch (factory.rs:376-383), selecting the latest via `max_by_key` and updating it in place rather than creating a duplicate. The only new test, `local_dev_nearai_mcp_auto_bootstraps_from_env_credentials` (factory.rs:3916), builds services exactly once, so it only exercises the fresh-account / `existing == None` path. The update branch — and the idempotency guarantee that a second startup does not produce a duplicate `nearai` account or a second activation — is never executed. This is the exact path the security focus flags (could re-bootstrap clobber/cross-bind a credential or accumulate duplicates). The module already has an established rebuild pattern (`local_dev_default_product_auth_preserves_manual_token_across_rebuilds`, factory.rs:3547) to copy.

**Fix:** Add tests::factory::local_dev_nearai_mcp_rebootstrap_updates_existing_account_without_duplication covering: build local-dev services with the env-credentials override active, build a second time against the same root with the override still active, then assert `accounts_for_owner` returns exactly one `nearai` account (no duplicate) whose access_secret is present/refreshed, and that the extension projection phase is still Active with a single active `nearai.web_search` capability. This drives the `with_update_binding` branch and the `LifecyclePhase::Active` no-op arm.

```rust
#[tokio::test]
async fn local_dev_nearai_mcp_rebootstrap_updates_existing_account_without_duplication() {
    let _override = crate::nearai_mcp::override_nearai_mcp_env_credentials_for_test(
        "https://private.near.ai", "nearai-test-key");
    let dir = tempfile::tempdir().expect("tempdir");
    let root = dir.path().join("local-dev");
    let owner = "nearai-rebootstrap-owner";
    build_reborn_services(RebornBuildInput::local_dev(owner, root.clone())).await.expect("first build");
    let services = build_reborn_services(RebornBuildInput::local_dev(owner, root.clone())).await.expect("second build");
    let scope = AuthProductScope::new(
        ResourceScope::local_default(UserId::new(owner).unwrap(), InvocationId::new()).unwrap(),
        AuthSurface::Api);
    let accounts = services.product_auth.as_ref().unwrap()
        .credential_account_record_source().accounts_for_owner(&scope).await.unwrap();
    let nearai: Vec<_> = accounts.iter().filter(|a| a.provider.as_str() == "nearai").collect();
    assert_eq!(nearai.len(), 1, "re-bootstrap must update in place, not duplicate");
    assert!(nearai[0].access_secret.is_some());
}
```

File: crates/ironclaw_reborn_composition/src/factory.rs [hunk 4]
```diff
@@ -320,6 +322,134 @@ where
...
+            reason: format!("NEAR AI MCP auto-enable scope could not be built: {error}"),
+        })?,
+        ironclaw_auth::AuthSurface::Api,
+    );
+    let provider = ironclaw_auth::AuthProviderId::new("nearai").map_err(|error| {
+        RebornBuildError::InvalidConfig {
+            reason: format!("NEAR AI MCP provider id is invalid: {error}"),
+        }
+    })?;
+    let existing = product_auth
```
Comment (abbyshekit) [inline crates/ironclaw_reborn_composition/src/factory.rs]: **[Medium · 80% confidence] Manual-token seeding reimplemented inline instead of reusing the canonical ProductAuthExtensionCredentialSetup helper**

_thermo reviewer_

The bootstrap reimplements the exact manual-token credential-seeding sequence that already exists as the canonical helper `ProductAuthExtensionCredentialSetup::submit_manual_token` in `crates/ironclaw_reborn_composition/src/webui_extension_credentials.rs` (lines ~65-97): build a `RebornManualTokenSetupRequest` with `AuthContinuationRef::SetupOnly` + a TTL, conditionally apply `with_update_binding(existing_account)`, call `request_manual_token_setup`, then `submit_manual_token(RebornManualTokenSubmitRequest::new(...))`. The bootstrap duplicates all of this (factory.rs 365-399) plus its own label construction and 300s/5-min TTL constant, and notably maps errors via raw `{error:?}` into a generic `RebornBuildError::InvalidConfig` instead of the sanitized `map_auth_error` the canonical helper uses. This is copy-pasted seeding logic in a second place, so the two flows will drift (TTL, label, error sanitization, future binding fields). The only genuinely bootstrap-specific step is resolving the existing account; that resolution can stay in the bootstrap while the request/submit goes through the shared service.

**Fix:** Resolve the existing account (the `accounts_for_owner` + `filter(provider) + max_by_key` block already present), then hand the `CredentialAccountUpdateBinding` plus scope/provider/label/secret to the existing `ProductAuthExtensionCredentialSetup::submit_manual_token` (constructed from the same `product_auth` Arc) rather than re-driving `request_manual_token_setup`/`submit_manual_token` inline. That deletes the duplicated setup-request construction, the duplicated TTL, and the divergent error path, and routes both seed callers through one sanitized helper. If `ExtensionCredentialSubmitRequest` does not currently carry a pre-resolved binding the way the bootstrap needs, extend that one request type rather than forking the flow here.

```rust
// after resolving `existing` -> Option<CredentialAccountUpdateBinding>
ProductAuthExtensionCredentialSetup::new(Arc::clone(product_auth))
    .submit_manual_token(ExtensionCredentialSubmitRequest {
        scope, provider, label: "NEAR AI API key".into(),
        secret: api_key, existing_account: existing_binding,
    })
    .await
    .map_err(/* map to RebornBuildError::InvalidConfig */)?;
```

File: crates/ironclaw_reborn_composition/src/factory.rs [hunk 5]
```diff
@@ -320,6 +322,134 @@ where
...
+            }
+        })?;
+    let phase = extension_management
+        .project(package_ref.clone())
+        .await
+        .map_err(|error| RebornBuildError::InvalidConfig {
+            reason: format!("NEAR AI MCP extension projection failed: {error}"),
+        })?
+        .phase;
+    match phase {
```
Comment (abbyshekit) [inline crates/ironclaw_reborn_composition/src/factory.rs]: **[Medium · 78% confidence] Duplicated activate() arm across Discovered/Installed phases — collapse to install-if-needed + single activate**

_thermo reviewer_

The phase `match` has two arms that both call `extension_management.activate(package_ref, ExtensionActivationMode::Static)` with byte-identical error mapping; the `Discovered` arm just prepends an `install`. `install` is not idempotent (extension_lifecycle.rs line ~648 errors `"... is already installed"`), so you genuinely cannot always-install — but the activate step is shared and copy-pasted. The duplication means a future change to the activation call (mode, error text, retry) must be made in two places, and the structure obscures the real invariant: "install only when Discovered, then always activate, unless already Active/Disabled."

**Fix:** Collapse the two auto-activatable arms into one branch: keep `Active => {}`, `Disabled => debug-log`, and `other => warn`, but merge `Discovered | Installed` into a single arm that conditionally installs (`if phase == Discovered { install()? }`) and then activates exactly once. This removes the duplicated activate+error-map block while preserving the install-precondition behavior.

```rust
match phase {
    LifecyclePhase::Discovered | LifecyclePhase::Installed => {
        if phase == LifecyclePhase::Discovered {
            extension_management.install(package_ref.clone()).await.map_err(install_err)?;
        }
        extension_management.activate(package_ref, ExtensionActivationMode::Static).await.map_err(activate_err)?;
    }
    LifecyclePhase::Active => {}
    LifecyclePhase::Disabled => tracing::debug!(/* preserve disabled */),
    other => tracing::warn!(phase = ?other, /* not auto-activatable */),
}
```

File: crates/ironclaw_reborn_composition/src/factory.rs [hunk 6]
```diff
@@ -320,6 +322,134 @@ where
...
+        LifecyclePhase::Installed => {
+            extension_management
+                .activate(package_ref, ExtensionActivationMode::Static)
+                .await
+                .map_err(|error| RebornBuildError::InvalidConfig {
+                    reason: format!("NEAR AI MCP extension activation failed: {error}"),
+                })?;
+        }
+        LifecyclePhase::Active => {}
+        LifecyclePhase::Disabled => {
```
Comment (abbyshekit) [inline crates/ironclaw_reborn_composition/src/factory.rs]: **[Medium · 70% confidence] disabled-state-preserved branch has no test despite being a documented guarantee**

_tests reviewer_

The `LifecyclePhase::Disabled` arm (factory.rs:438-442) is the mechanism behind the explicit promise in docs/reborn/onboarding.md: "users can disable NEAR AI MCP after bootstrap and startup will not re-enable it." No test exercises this branch. If a future change reordered the match or fell through `Disabled` into an activate call, the regression would ship silently and violate the documented contract. The single happy-path test only hits `Discovered`; no test puts the extension into `Disabled` and re-runs the bootstrap to confirm it stays disabled.

**Fix:** Add tests::factory::local_dev_nearai_mcp_bootstrap_preserves_explicit_disabled_state: build local-dev services (extension becomes Active), disable the `nearai` extension via the extension_management port, build services again with the env override still active, then assert the projected phase is still `LifecyclePhase::Disabled` and that `nearai.web_search` is absent from active_model_visible_capabilities().

```rust
// after disabling the nearai extension, rebuild and assert it stays disabled:
let projection = extension_management.project(nearai_ref).await.expect("projected");
assert_eq!(projection.phase, LifecyclePhase::Disabled);
```

File: crates/ironclaw_reborn_composition/src/factory.rs (hunk 2 above)
Comment (abbyshekit) [inline crates/ironclaw_reborn_composition/src/factory.rs]: **[Low · 70% confidence] 126-line NEAR-AI-MCP bootstrap added to an already-4671-line factory.rs; belongs in the nearai_mcp sibling module**

_thermo reviewer_

`bootstrap_local_dev_nearai_mcp_from_env` (factory.rs 325-451, ~126 lines) is a cohesive, single-purpose routine that lives in a file that is already 4671 lines and growing. The entire concern it implements — read NEARAI_BASE_URL/API_KEY, validate the endpoint, seed the credential, install+activate the bundled extension — is the same concern owned by the sibling module `crates/ironclaw_reborn_composition/src/nearai_mcp.rs`, which it already calls into twice (`nearai_mcp_env_credentials`, `nearai_mcp_endpoint_from_base`). `nearai_mcp.rs` is a pure leaf module (only imports `std::net::IpAddr`) and all the types the bootstrap touches (`RebornProductAuthServices`, `RebornLocalExtensionManagementPort`, `RebornManualTokenSetupRequest`, `LifecyclePackageRef`, `RebornBuildError`) are crate-internal, so the move adds imports but introduces no cycle. factory.rs's call site shrinks to a one-line delegation.

**Fix:** Move `bootstrap_local_dev_nearai_mcp_from_env` (and the test override plumbing it pairs with) into `nearai_mcp.rs` as `pub(crate) async fn bootstrap_local_dev_nearai_mcp_from_env(...)`, keeping the single call in `build_local_dev` as `crate::nearai_mcp::bootstrap_local_dev_nearai_mcp_from_env(&product_auth, &extension_management, &owner).await?`. This co-locates all NEAR-AI-MCP env-bootstrap logic in one module and stops factory.rs accreting feature-specific routines.

```rust
// factory.rs build_local_dev:
crate::nearai_mcp::bootstrap_local_dev_nearai_mcp_from_env(
    &product_auth, &extension_management, &owner_user_id_for_nearai_mcp,
).await?;
```

File: crates/ironclaw_reborn_composition/src/factory.rs [hunk 7]
```diff
@@ -320,6 +322,134 @@ where
...
 
+async fn bootstrap_local_dev_nearai_mcp_from_env(
+    product_auth: &Arc<RebornProductAuthServices>,
+    extension_management: &Arc<RebornLocalExtensionManagementPort>,
+    owner_user_id: &UserId,
+) -> Result<(), RebornBuildError> {
+    let Some((base_url, api_key)) = crate::nearai_mcp::nearai_mcp_env_credentials() else {
+        return Ok(());
+    };
+    crate::nearai_mcp::nearai_mcp_endpoint_from_base(Some(&base_url)).map_err(|error| {
```
Comment (abbyshekit) [inline crates/ironclaw_reborn_composition/src/factory.rs]: **[Low · 60% confidence] invalid NEARAI_BASE_URL -> build failure error path is not exercised end-to-end**

_tests reviewer_

`bootstrap_local_dev_nearai_mcp_from_env` maps a rejected base URL from `nearai_mcp_endpoint_from_base` into `RebornBuildError::InvalidConfig`, which fails the entire local-dev build (factory.rs:333-337). The URL-validation rejections themselves are unit-tested in nearai_mcp.rs::tests (`endpoint_validation_rejects_http_and_forbidden_ips`), but the integration — a misconfigured/attacker-controlled NEARAI_BASE_URL causing `build_reborn_services` to return Err rather than silently activating the MCP against a bad host — has no test. The override helper accepts an arbitrary base_url string, so this path is now testable.

**Fix:** Add tests::factory::local_dev_nearai_mcp_bootstrap_rejects_invalid_base_url: set the env override to a forbidden base (e.g. "http://169.254.169.254" or "https://169.254.169.254"), call build_reborn_services(local_dev(..)).await and assert it returns Err(RebornBuildError::InvalidConfig { .. }) with a reason mentioning NEARAI_BASE_URL, confirming the bootstrap fails closed instead of activating against a disallowed endpoint.

```rust
let _o = crate::nearai_mcp::override_nearai_mcp_env_credentials_for_test("https://169.254.169.254", "k");
let err = build_reborn_services(RebornBuildInput::local_dev("o", root)).await.expect_err("must fail closed");
assert!(matches!(err, RebornBuildError::InvalidConfig { .. }));
```

File: crates/ironclaw_reborn_composition/src/nearai_mcp.rs [hunk 8]
```diff
@@ -15,6 +15,62 @@ pub(crate) fn nearai_mcp_endpoint_from_env() -> Result<NearAiMcpEndpoint, String
     nearai_mcp_endpoint_from_base(configured_base.as_deref())
 }
 
+pub(crate) fn nearai_mcp_env_credentials() -> Option<(String, secrecy::SecretString)> {
```
Comment (abbyshekit) [inline crates/ironclaw_reborn_composition/src/nearai_mcp.rs]: **[Low · 55% confidence] nearai_mcp_env_credentials one-var-missing / empty / whitespace branches are unexercised**

_tests reviewer_

The new `nearai_mcp_env_credentials` (nearai_mcp.rs:18-33) trims both env values and early-returns None if either is missing or empty, returning Some only when both are non-empty. The test-only override short-circuits this and always returns a full pair, so the production trim + is_empty filter + one-var-missing fallthrough is never executed by any test. The security/bugs focus calls out exactly these edge cases (one of two vars set, empty/whitespace). Note: directly testing the real `std::env::var` path is racy due to global env state and is partly shadowed by the `#[cfg(test)]` override, so a parser-shaped helper is the clean target.

**Fix:** Extract the trim/non-empty pairing into a pure helper, e.g. fn nearai_mcp_credentials_from_parts(base: Option<String>, key: Option<String>) -> Option<(String, SecretString)>, and add tests::nearai_mcp::credentials_require_both_nonempty_trimmed covering: both set -> Some (values trimmed); base set / key missing -> None; key set / base missing -> None; "  " whitespace-only -> None; leading/trailing whitespace -> trimmed Some. nearai_mcp_env_credentials then calls the helper with the two env reads.

```rust
#[test]
fn credentials_require_both_nonempty_trimmed() {
    assert!(nearai_mcp_credentials_from_parts(Some(" https://x ".into()), None).is_none());
    assert!(nearai_mcp_credentials_from_parts(None, Some("k".into())).is_none());
    assert!(nearai_mcp_credentials_from_parts(Some("   ".into()), Some("k".into())).is_none());
    let got = nearai_mcp_credentials_from_parts(Some(" https://x ".into()), Some(" k ".into())).unwrap();
    assert_eq!(got.0, "https://x");
}
```

File: crates/ironclaw_reborn_composition/src/factory.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (hanakannzashi) [reply crates/ironclaw_reborn_composition/src/factory.rs]: Thanks for flagging this. I am leaving this one unresolved intentionally rather than folding a lifecycle atomicity refactor into this PR.

The current local-dev bootstrap no longer owns the NEAR AI state machine in `factory.rs`; it delegates through the lifecycle port, whose mutating operations take the lifecycle operation lock and fail closed if the state changes underneath them. That means a concurrent change can still make the bootstrap return an error / skip activation, but it should not silently expose partial Reborn traffic or bypass lifecycle policy.

A truly atomic `project + install-if-needed + activate` operation would need a new lifecycle-port helper with lock-held internal install/activate steps, plus concurrency coverage around install/activate/remove and hosted-MCP discovery. That is a broader lifecycle hardening slice than this NEAR AI MCP config/bootstrap PR. I think it is better tracked as follow-up work instead of making this PR carry an extension lifecycle API refactor.

**Review by hanakannzashi — APPROVED** [review]
